src-built-in/components/UIAPI/reducers/linker.ts

```diff
@@ -0,0 +1,184 @@
+            nextChannelId += 1;
+        });
+        linkerInitState.channels = initialChannels;
+        linkerInitState.nameToId = initialNametoId;
+
+        FSBL.Clients.RouterClient.addResponder("Finsemble.LinkerWindow.SetActiveChannels", function (err: any, msg: any) {
+            if (err) {
+                return rej(`Failed to add Finsemble.LinkerWindow.SetActiveChannels Responder: ${err}`);
+            }
+            const activeChannels = msg.data.channels;
+
+            const activeChannelIds: any[number] = [];
+            activeChannels.forEach((channel: Channel) => {
+                activeChannelIds.push(linkerInitState.nameToId[channel.name]);
+            });
+            const updatedChannels: Channels = Object.assign({}, linkerInitState.channels);
+            activeChannelIds.forEach((channelId: number) => {
+                updatedChannels[channelId].active = true;
+            });
+
+            linkerInitState.windowIdentifier = msg.data.windowIdentifier;
+            FSBL.Clients.Logger.system.log("toggle Linker window");
+            msg.sendQueryResponse(null, {});
+            store.dispatch(updateActiveChannels(msg.data));
+            
+            FSBL.Clients.ConfigClient.getValue("finsemble.accessibleLinker", (err: any, value: boolean) => {
+                if (err) {
+                    rej(`Error getting accessibleLinker value: ${err}`);
+                }
+                const newLinkerState = {
+                    ...linkerInitState,
+                    channels: updatedChannels,
+                    isAccessibleLinker: value,
+                };
+                res(newLinkerState);
+            });
+        });
+    });
+}
+
+function cleanUpAfterComponentUnmount() {
+    finsembleWindow.removeEventListener("blurred", () => {
+        finsembleWindow.hide();
+    });
+    finsembleWindow.removeEventListener("shown", () => {
+		finsembleWindow.focus();
+	});
+}
+
+// The linker's reducer
+const linker = (state = initialState, action: LinkerAction) => {
+    const { type, payload } = action;
+    switch (type) {
+        case LINKER_INIT:            
+            return loop(state, Cmd.run(initializeLinker, {
+                successActionCreator: initSuccess,
+            }));
+        case LINKER_INIT_SUCCESS:
+            const newState = payload.value;
+            return loop(newState, Cmd.run(() => FSBL.Clients.WindowClient.fitToDOM()));
+        case TOGGLE_CHANNEL_REQUEST:
+            const newState_request: Linker = {
+                ...state,
+                processingRequest: true
+            };
+
+            const cmd = Cmd.run(linkChannel, {
+                successActionCreator: () => toggleSuccess(payload.channelID),
+                failActionCreator: () => toggleFailure(),
+                args: [newState_request.channels[payload.channelID].name, newState_request.channels[payload.channelID].active, newState_request.windowIdentifier]
```

Comment:
i really don't like these args. Please declare variables instead of accessing all of these values. What if newState_request.channels === undefined? All this stuff would barf.

Reply:
Is it [realistically] possible for newState_request.channels to be undefined though? A developer would need to write code to actively delete the channels member. That seems highly unlikely in unidirectional flow.

Reply:
The state would be initialized when the component first mounts and we call `linkerActions.init()`. If the application ever gets to this code it means we have already successfully initialized the linker. However, we do need to discuss what we should do with the init failure case, for now it does nothing other than some logging messages. The linker code before the refactor didn't address this and I'm simply modeling the previous behavior.